deepface/modules/recognition.py

```diff
@@ -415,3 +440,219 @@ def __find_bulk_embeddings(
+        List[List[Dict[str, Any]]]: 
+            A list where each element corresponds to a source face and
+            contains a list of dictionaries with matching faces.
+    """
+    embeddings_list = []
+    valid_mask = []
+    other_keys = set()
+
+    for item in representations:
+        emb = item.get('embedding')
+        if emb is not None:
+            embeddings_list.append(emb)
+            valid_mask.append(True)
+        else:
+            embeddings_list.append(np.zeros_like(representations[0]['embedding']))
+            valid_mask.append(False)
+
+        other_keys.update(item.keys())
+
+    # remove embedding key from other keys
+    other_keys.discard('embedding')
+    other_keys = list(other_keys)
+
+    embeddings = np.array(embeddings_list) # (N, D)
+    valid_mask = np.array(valid_mask) # (N,)
+
+    data = {
+        key: np.array([item.get(key, None) for item in representations])
+        for key in other_keys
+    }
+
+    target_embeddings = []
+    source_regions = []
+    target_thresholds = []
+
+    for source_obj in source_objs:
+        if anti_spoofing and not source_obj.get("is_real", True):
+            raise ValueError("Spoof detected in the given image.")
+
+        source_img = source_obj["face"]
+        source_region = source_obj["facial_area"]
+
+        target_embedding_obj = representation.represent(
+            img_path=source_img,
+            model_name=model_name,
+            enforce_detection=enforce_detection,
+            detector_backend="skip",
+            align=align,
+            normalization=normalization,
+        )
+        # it is safe to access 0 index because we already fed detected face to represent function
+        target_representation = target_embedding_obj[0]["embedding"]
+
+        target_embeddings.append(target_representation)
+        source_regions.append(source_region)
+
+        target_threshold = threshold or verification.find_threshold(model_name, distance_metric)
+        target_thresholds.append(target_threshold)
+
+    target_embeddings = np.array(target_embeddings) # (M, D)
+    target_thresholds = np.array(target_thresholds) # (M,)
+    source_regions_arr = {
+        'source_x': np.array([region['x'] for region in source_regions]),
+        'source_y': np.array([region['y'] for region in source_regions]),
+        'source_w': np.array([region['w'] for region in source_regions]),
+        'source_h': np.array([region['h'] for region in source_regions]),
+    }
+
+    def find_cosine_distance_batch(
+        embeddings: np.ndarray, target_embeddings: np.ndarray
+    ) -> np.ndarray:
+        """
+        Find the cosine distances between batches of embeddings
+        Args:
+            embeddings (np.ndarray): array of shape (N, D)
+            target_embeddings (np.ndarray): array of shape (M, D)
+        Returns:
+            np.ndarray: distance matrix of shape (M, N)
+        """
+        embeddings_norm = verification.l2_normalize(embeddings, axis=1)
```

Comment:
you performed cosine l2 here but we mostly don't apply l2 to cosine

Reply:
L2 normalization is exactly what `np.linalg.norm` do. The `l2_normalize` method is just a wrap around `np.linalg.norm`. Therefore applying `verification.l2_normalize` in the `find_cosine_distance_batch` is the same as applying `np.linalg.norm` in the `find_cosine_distance`, but taking into account that we do it batch-wise:

```python
def find_cosine_distance(
    source_representation: Union[np.ndarray, list], test_representation: Union[np.ndarray, list]
) -> np.float64:
    ...
    a = np.dot(source_representation, test_representation)
    b = np.linalg.norm(source_representation)
    c = np.linalg.norm(test_representation)
    return 1 - a / (b * c)
```

```python
def find_cosine_distance_batch(
    embeddings: np.ndarray, target_embeddings: np.ndarray
) -> np.ndarray:
    ...
    embeddings_norm = verification.l2_normalize(embeddings, axis=1)
    target_embeddings_norm = verification.l2_normalize(target_embeddings, axis=1)
    cosine_similarities = np.dot(target_embeddings_norm, embeddings_norm.T)
    cosine_distances = 1 - cosine_similarities
    return cosine_distances
```

Example to test:
```python
x = np.random.randn(100)
y = np.random.randn(100)
ans1 = find_cosine_distance(x, y)
ans2 = find_cosine_distance_batch([x], [y])
print(np.round(ans1, 10) == np.round(ans2, 10)) # True
```